src/Database/PostgreSQL/PQTypes/Model/Migration.hs

```diff
@@ -57,6 +59,17 @@ data MigrationAction m =
       TableIndex
 #endif
 
+  -- | Migration for modifying columns. Parameters are:
+  --
+  -- SQL that will be used for the cursor.
+  --
+  -- Function that takes a list of primary keys provided by the cursor SQL and
+  -- runs an arbitrary computation within MonadDB. The function might be called
+  -- repeatedly depending on the number of primary keys. See the last argument.
+  --
+  -- Number of primary keys fetched at once by the cursor SQL.
+  | forall t . FromRow t => ModifyColumnMigration SQL ([t] -> m ()) Int
```

Comment:
In principle this is correct, but it really should be `FromSQL`, not `FromRow`. The problem is multi-column primary key support - we can fetch multi-column primary keys, but there is no way to feed them back to Postgres as that would require us to be able to pass arrays of tuples as a query parameter, which the library doesn't support.

This is not the end of the world as doing migrations for such tables is still possible, it just needs to use the most reasonable single column part of the primary key. By the most reasonable I mean such that the amount of rows handled in a single batch will be similar to the amount of rows that would be handled if the full primary key was used.

For example, if for the table `author_attachments` the primary key is `(document_id, file_id, name)`, using the `document_id` part is fine, let's say with the batch of 500 (if 1000 would be used by default) with the assumption that the document has 2 attachments on average.

Of course that's a complication that needs to be documented and understood by people, so I'm wondering whether using a temporary table might not be better in the end :thinking: :disappointed:

Reply:
Nevermind, `FromRow` is fine. Turns out that handling multi-column priary keys is possible, just slightly involved:

1. You get the list of tuples from PostgreSQL.
2. You unzip them into a tuple of lists in Haskell.
3. You pass the lists to PostgreSQL as separate parameters and zip them back in the SQL, see https://stackoverflow.com/questions/12414750/is-there-something-like-a-zip-function-in-postgresql-that-combines-two-arrays for more details.

Let's add that to the Haddock so people know how to handle multi-column primary keys.